ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -0,0 +1,313 @@
+    {
+      displayChildren(m_kmlNodesList[0]);
+    }
+  });
+}
+
+QString ListKmlContents::getKmlNodeType(KmlNode *node)
+{
+  if (node == nullptr)
+    return "";
+
+  QString type = "";
+  switch (node->kmlNodeType()) {
+  case KmlNodeType::KmlDocument:
+    type = "KmlDocument";
+    break;
+  case KmlNodeType::KmlFolder:
+    type = "KmlFolder";
+    break;
+  case KmlNodeType::KmlNetworkLink:
+    type = "KmlNetworkLink";
+    break;
+  case KmlNodeType::KmlPlacemark:
+    type = "KmlPlacemark";
+    break;
+  case KmlNodeType::KmlPhotoOverlay:
+    type = "KmlPhotoOverlay";
+    break;
+  case KmlNodeType::KmlGroundOverlay:
+    type = "KmlGroundOverlay";
+    break;
+  case KmlNodeType::KmlScreenOverlay:
+    type = "KmlScreenOverlay";
+    break;
+  case KmlNodeType::KmlTour:
+    type = "KmlTour";
+    break;
+  default:
+    return "";
+  }
+  return type;
+}
+
+void ListKmlContents::displayChildren(KmlNode* parentNode)
+{
+  if (parentNode == nullptr)
+    return;
+
+  if (KmlContainer* container = dynamic_cast<KmlContainer*>(parentNode))
+  {
+    m_levelNodeNames.clear();
+    bool lastLevel = true;
+
+    // for current level, get names of child nodes
+    for (KmlNode* node: *(container->childNodesListModel()))
+    {
+      QString str = node->name() + " - " + getKmlNodeType(node);
+
+      // if node has children, add ">" to indicate further levels
+      if (!node->children().isEmpty())
+      {
+        str.append(" >");
+        lastLevel = false;
+      }
+      m_levelNodeNames << str;
+    }
+
+    if (lastLevel)
+    {
+      m_currentNode = container->childNodesListModel()->at(0);
+      emit currentNodeChanged();
+    }
+    emit levelNodeNamesChanged();
+  }
+}
+
+// recursively build string to indicate node's ancestors
+void ListKmlContents::buildPathLabel(KmlNode* node)
+{
+  if (node == nullptr)
+    return;
+
+  if (node->parentNode() != nullptr)
+  {
+    buildPathLabel(node->parentNode());
+    m_labelText.append(">");
+  }
+  m_labelText.append(node->name());
+}
+
+void ListKmlContents::displayPreviousLevel()
+{
+  KmlNode* parentNode = m_currentNode->parentNode();
+  if (parentNode == nullptr)
+    return;
+
+  KmlNode* grandparentNode = parentNode->parentNode();
+
+  if (grandparentNode != nullptr)
+  {
+    displayChildren(grandparentNode);
+    m_currentNode = grandparentNode;
+    emit currentNodeChanged();
+    emit labelTextChanged();
+  }
+  // if grandparent node is nullptr, then at top of tree
+  else
+  {
+    displayChildren(parentNode);
+    m_currentNode = parentNode;
+    emit currentNodeChanged();
+    emit labelTextChanged();
+  }
+}
+
+// display selected node on sceneview and show its children
+void ListKmlContents::processSelectedNode(const QString& nodeName)
+{
+  QString extractedNodeName = nodeName;
+  if (nodeName.contains(" - "))
+  {
+    int ind = nodeName.lastIndexOf(" - ");
+    extractedNodeName = nodeName.left(ind);
+  }
```

Comment:
It took me a while to realize what was going on here. 

You don't necessarilly need to do any string manipulation to match the names.

You could just do:

```cpp
 if (nodeName.startsWith(node->name()))
 {
...
```

Reply:
I don't think this will work in cases where 2 names begin in the same way. For example, we have "Building O" and "Building OA" as 2 different nodes

Reply:
Ahh good call